### PR TITLE
Add top level domains

### PR DESCRIPTION
**Pull request by PrinsFrank**

### CODE REVIEW

File: dev/DataSource/Mapping/TopLevelDomainMapping.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        foreach ($items as $item) {
+            $columns = $item->findElements(WebDriverBy::xpath('./td'));
+            if (count($columns) !== 3) {
+                continue;
+            }
+
+            $record          = (object) [];
+            $record->tld     = $columns[0]->getText();
+            $record->type    = $columns[1]->getText();
+            $record->manager = $columns[2]->getText();
```
Comment (szepeviktor) [inline dev/DataSource/Mapping/TopLevelDomainMapping.php:48]: There are 139 revoked TLD-s. "Not assigned"

File: dev/DataSource/Mapping/TopLevelDomainMapping.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        $items = $crawler->filterXPath('//table[@id="tld-table"]/tbody/tr')->getIterator();
+
+        $dataSet = [];
+        foreach ($items as $item) {
+            $columns = $item->findElements(WebDriverBy::xpath('./td'));
+            if (count($columns) !== 3) {
+                continue;
+            }
+
+            $record          = (object) [];
```
Comment (szepeviktor) [inline dev/DataSource/Mapping/TopLevelDomainMapping.php:45]: Without hacking it is really
```suggestion
            $record          = new stdClass();
```

File: src/TopLevelDomain/CountryCodeTLD.php [hunk 3]
```diff
@@ -5,4 +5,305 @@
 
 enum CountryCodeTLD: string
 {
+    case  = 'ලංකා';
```
Comment (szepeviktor) [inline src/TopLevelDomain/CountryCodeTLD.php:8]: We should throw an exception on empty keys.

Comment (PrinsFrank) [reply src/TopLevelDomain/CountryCodeTLD.php]: Agreed! You can create a PR if you want? Otherwise I'll create one. ;)

Comment (PrinsFrank) [reply src/TopLevelDomain/CountryCodeTLD.php]: I'll need to dive in the underlying issue where Singhalan is not transliterated at all.

Comment (szepeviktor) [reply src/TopLevelDomain/CountryCodeTLD.php]: 👀 https://unicode.org/charts/PDF/U0D80.pdf

File: dev/DataTarget/NameNormalizer.php [hunk 4]
```diff
@@ -54,9 +55,8 @@ public static function normalize(string $key): string
         }
 
         $key = trim(str_replace(['__', '__'], ['_', '_'], $key), '_');
-
         if ($key === '') {
-            throw new TransliterationException('No characters left in key');
+            throw new TransliterationException(sprintf('No characters left in key (%s)', $key));
```
Comment (szepeviktor) [inline dev/DataTarget/NameNormalizer.php:59]: @PrinsFrank My LASER eyes see an empty string here 🙃

File: src/TopLevelDomain/CountryCodeTLD.php (hunk 3 above)
Comment (szepeviktor) [reply src/TopLevelDomain/CountryCodeTLD.php]: image

https://aksharamukha.appspot.com/converter

File: dev/DataSource/Mapping/TopLevelDomainMapping.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            if (count($columns) !== 3) {
+                continue;
+            }
+
+            $record = (object) [];
+            $record->tld = $columns[0]->getText();
+            $record->type = $columns[1]->getText();
+            $record->manager = $columns[2]->getText();
+
+            /** @var TDataSet $record */
```
Comment (szepeviktor) [inline dev/DataSource/Mapping/TopLevelDomainMapping.php:50]:
```suggestion
            // Exclude revoked TLD-s.
            if ($record->manager === 'Not assigned') {
                continue;
            }

            /** @var TDataSet $record */
```

Comment (PrinsFrank) [reply dev/DataSource/Mapping/TopLevelDomainMapping.php:55]: Thanks, committed!

**Comment by szepeviktor** [comment]
>   No characters left in key (ලංකා)  

That reads "lankaa"

File: dev/DataTarget/NameNormalizer.php [hunk 6]
```diff
@@ -54,6 +55,8 @@ public static function normalize(string $key): string
             ->replace('|', '_')
             ->replace('}', '_')
             ->replace('~', '_')
+            ->replace('ලංකා', 'lanka')
+            ->replace('ລາວ', 'laav')
```
Comment (szepeviktor) [inline dev/DataTarget/NameNormalizer.php:59]: "ລາວ" is in Lao. It is transliterated as "lao".

Comment (szepeviktor) [reply dev/DataTarget/NameNormalizer.php:59]: https://unicode.org/charts/PDF/U0E80.pdf

**Comment by szepeviktor** [comment]
What happened to sorting? Has it changed from key-based to value-based?

**Comment by szepeviktor** [comment]
> Has it changed from key-based to value-based?

No. Enums are sorted by the **original** name (before transliteration): "Åland Islands"

**Comment by PrinsFrank** [comment]
@szepeviktor Fixed!

**Review by szepeviktor — APPROVED** [review]

**Pull request merged**